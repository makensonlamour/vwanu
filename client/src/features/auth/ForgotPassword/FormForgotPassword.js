import React from "react";
import { v4 as uuidv4 } from "uuid";
import * as Yup from "yup";
import { useSelector, useDispatch } from "react-redux";
import { Link, useNavigate } from "react-router-dom";
import routesPath from "../../../routesPath";

//RTK query
import { useForgotPasswordMutation } from "../../api/apiSlice";
import { getAlerts, setAlert, removeAlert } from "../../alert/alertSlice";

// Core components
import Alerts from "../../../components/common/Alerts";
import { Field, Form, Submit } from "../../../components/form";
import Loader from "../../../components/common/Loader";

const ValidationSchema = Yup.object().shape({
  email: Yup.string().required().min(6).email().label("Email"),
});

const initialValues = {
  email: "",
};

const FormForgotPassword = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const alerts = useSelector(getAlerts);
  const [forgotPassword, { isLoading }] = useForgotPasswordMutation();
  const id = uuidv4();

  const handleResetPasword = async (credentials) => {
    try {
      const data = await forgotPassword(credentials).unwrap();
      navigate(routesPath.FORGOT_PASSWORD_SUCCESS, { state: data, email: credentials.email });
    } catch (e) {
      console.log();
      let customMessage = "An unknown network error has occurred on Vwanu. Try again later.";
      if (e.status === 400) {
        dispatch(
          setAlert({
            msg: e.data.errors[0].message,
            id,
            type: "error",
          })
        );

        // Setting a sec for each 10 words
        setTimeout(() => dispatch(removeAlert(id)), e.data.errors[0].message.length * 200);
      } else {
        dispatch(
          setAlert({
            msg: customMessage,
            id,
            type: "error",
          })
        );

        // Setting a sec for each 10 words
        setTimeout(() => dispatch(removeAlert(id)), customMessage.length * 200);
      }
    }
  };
  return (
    <>
      <Form
        validationSchema={ValidationSchema}
        initialValues={initialValues}
        onSubmit={handleResetPasword}
        className="shadow-lg rounded-3xl my-10"
      >
        <h1 className="card-title text-secondary text-md text-center">Forgot your password?</h1>
        <p className="text-sky-600">{`Enter your email and we'll send you a link to reset your password.`}</p>
        <Alerts className="bg-error mt-4" alerts={alerts} />
        <Field
          required
          autoCapitalize="none"
          placeholder="Email"
          name="email"
          type="email"
          autoComplete="new-email"
          className="mt-1 lg:mt-2 bg-blue-200 text-secondary placeholder:text-secondary font-semibold rounded-full input-secondary border-none invalid:text-red-600 autofill:text-secondary autofill:bg-blue-200"
          testId="email-error-message"
        />

        <Submit data-testid="forgotPasswordBtn" className="rounded-full text-md" title={isLoading ? <Loader /> : "Reset Password"} />
        <div className="divider">OR</div>
        <Link className="text-primary font-bold mb-10 text-center" to={routesPath.REGISTER}>
          Create New Account
        </Link>
        <Link className="text-primary font-bold mt-10 text-center" to={routesPath.LOGIN}>
          Back to Login
        </Link>
      </Form>
    </>
  );
};

export default FormForgotPassword;
