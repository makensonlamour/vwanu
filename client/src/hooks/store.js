import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../features/auth/authSlice";
import alertReducer from "../features/alert/alertSlice";
import userReducer from "../features/user/userSlice";
import { apiSlice } from "../features/api/apiSlice";

export const store = configureStore({
  reducer: {
    user: userReducer,
    alert: alertReducer,
    auth: authReducer,
    [apiSlice.reducerPath]: apiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiSlice.middleware),
});
