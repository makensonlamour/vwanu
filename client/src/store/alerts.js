import { createSlice } from "@reduxjs/toolkit";
const initialState = {
  list: [],
};
const Alert = createSlice({
  name: "alerts",
  initialState,
  reducers: {
    setAlert: (state, action) => {
      state.list.push(action.payload);
    },
    removeAlert: (state, action) => {
      const firstIndex = state.list.findIndex((alert) => alert.id === action.payload);

      state.list.splice(firstIndex, 1);
    },
  },
});

export default Alert.reducer;
export const { removeAlert, setAlert } = Alert.actions;
export const getAlerts = (state) => state.alerts;
