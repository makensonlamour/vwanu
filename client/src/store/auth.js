import * as action from "./api";
import env from "../config";
import { createSlice } from "@reduxjs/toolkit";
import jwtDecode from "jwt-decode";
import { saveToken, deleteToken } from "../middleware/api";

const url = env.endpoints;
//const { EXPECTED_HEADER } = env;

const initialState = {
  loading: false,
  data: null,
  lastFetch: null,
  token: null,
  error: null,
};

export const Auth = createSlice({
  name: "auth",
  initialState,
  reducers: {
    loginRequested: (state) => {
      state.loading = true;
    },
    loginSucceed: (state, action) => {
      state.loading = false;
      state.token = action.payload.data.token;
      saveToken(action.payload.data.token);
      //state.data = action.payload.data.user;
      state.data = jwtDecode(action.payload.data.token);
      state.lastFetch = Date.now;
    },

    LoginFailed: (state, action) => {
      state.loading = false;
      state.token = null;
      state.error = action.payload;
      state.data = null;
      deleteToken();
    },
    setUser: (state, action) => {
      state.loading = false;
      state.token = action.payload;
      saveToken(action.payload);
      state.data = jwtDecode(action.payload);
    },
    LogOut: (state) => {
      state.loading = false;
      state.token = null;
      state.data = null;
      state.error = null;
      deleteToken();
      window.location.replace("/login");
    },
  },
});

export const Login = (credentials) => (dispatch, getState) => {
  const { data, loading } = getState();
  if (data != null || loading) return;
  dispatch(
    action.apiCallBegan({
      url: url.LOGIN,
      data: credentials,
      method: "POST",
      onSuccess: Auth.actions.loginSucceed.type,
      onStart: Auth.actions.loginRequested.type,
      onError: Auth.actions.LoginFailed.type,
    })
  );
};

export const createUser = (newUserData) => (dispatch) => {
  dispatch(
    action.apiCallBegan({
      url: url.REGISTER,
      data: newUserData,
      method: "POST",
      onSuccess: Auth.actions.loginSucceed.type,
      onStart: Auth.actions.loginRequested.type,
      onError: Auth.actions.LoginFailed.type,
    })
  );
};

export const forgotPassword = (newUserData) => (dispatch) => {
  dispatch(
    action.apiCallBegan({
      url: url.FORGOT_PASSWORD,
      data: newUserData,
      method: "POST",
      onSuccess: Auth.actions.loginSucceed.type,
      onStart: Auth.actions.loginRequested.type,
      onError: Auth.actions.LoginFailed.type,
    })
  );
};

export const resetPassword = (newUserData) => (dispatch) => {
  dispatch(
    action.apiCallBegan({
      url: url.RESET_PASSWORD,
      data: newUserData,
      method: "POST",
      onSuccess: Auth.actions.loginSucceed.type,
      onStart: Auth.actions.loginRequested.type,
      onError: Auth.actions.LoginFailed.type,
    })
  );
};

export const { setUser } = Auth.actions;
export const logout = () => (dispatch) => dispatch(Auth.actions.LogOut());
export const getCurrentUser = (state) => state.authentication;
export const logged = Auth.actions.loginSucceed.type;
export default Auth.reducer;
