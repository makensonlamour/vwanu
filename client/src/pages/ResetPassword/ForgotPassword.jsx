import React from "react";

// Core components
import FormForgotPassword from "../../features/auth/ForgotPassword/FormForgotPassword";
import logo_mobile from "../../assets/images/Asset_2.png";

const ForgotPasswordScreen = () => {
  return (
    <>
      <div className="">
        <div className="mt-10 mx-10">
          {/*logo for responsive mobile*/}
          <div className="mt-4 mb-6 lg:mb-10">
            <img className="w-2/5 lg:w-1/5 m-auto" src={logo_mobile} alt="logo_vwanu" />
          </div>
          <div className="justify-center m-auto md:px-16 lg:px-0 md:w-2/3 lg:w-1/3">
            <FormForgotPassword />
          </div>
        </div>
      </div>
    </>
  );
};

export default ForgotPasswordScreen;
