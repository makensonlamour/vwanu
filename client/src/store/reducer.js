import { combineReducers } from "redux";

import alerts from "./alerts";
import authentication from "./auth";

export default combineReducers({
  authentication,
  alerts,
});
