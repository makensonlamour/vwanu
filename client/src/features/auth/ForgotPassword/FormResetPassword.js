import React, { useState } from "react";
import * as Yup from "yup";
import { Link, useNavigate, useParams } from "react-router-dom";
import routesPath from "../../../routesPath";
import { useResetPassword } from "../../auth/authSlice";
import { alertService } from "../../../components/common/Alert/Services";
import { Alert } from "../../../components/common/Alert";
import { Field, Form, Submit } from "../../../components/form";
import Loader from "../../../components/common/Loader";

const ValidationSchema = Yup.object().shape({
  password: Yup.string().required().min(8).label("Password"),
  passwordConfirmation: Yup.string()
    .required()
    .oneOf([Yup.ref("password"), null], "Passwords must be match"),
});

const FormResetPassword = () => {
  const { idUser, resetPasswordKey } = useParams();
  const navigate = useNavigate();
  const [isLoading, setIsLoading] = useState(false);

  const initialValues = {
    password: "",
    passwordConfirmation: "",
    idUser,
    resetPasswordKey,
  };

  const resetPassword = useResetPassword(["reset", "password"], { idUser, resetPasswordKey }, undefined, undefined);

  const handleResetPasword = async (credentials) => {
    setIsLoading(true);
    try {
      const data = await resetPassword.mutateAsync({
        action: "resetPwdLong",
        value: { token: resetPasswordKey, password: credentials?.password },
      });
      if (data) {
        navigate(routesPath.LOGIN);
      }
    } catch (e) {
      if (e?.response?.status === 400) {
        alertService.error(e?.response?.data?.errors[0]?.message, { autoClose: true });
      } else {
        alertService.error("An unknown network error has occurred on Vwanu. Try again later.", { autoClose: true });
      }
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <>
      <Form
        validationSchema={ValidationSchema}
        initialValues={initialValues}
        onSubmit={handleResetPasword}
        className="shadow-lg rounded-3xl my-10"
      >
        <h1 className="card-title text-secondary text-md text-center">Create New password</h1>
        <Alert />
        <Field
          required
          autoCapitalize="none"
          autoCorrect="false"
          placeholder="New Password"
          name="password"
          autoComplete="new-email"
          containerClassName="my-4"
          className="mr-1 mt-1 lg:mt-2 bg-blue-200 text-secondary placeholder:text-secondary font-semibold rounded-full input-secondary border-none invalid:text-red-500 autofill:text-secondary autofill:bg-blue-200"
          testId="password-error-message"
          showPassword={true}
        />

        <Field
          required
          autoCapitalize="none"
          autoCorrect="false"
          placeholder="Password Confirmation"
          name="passwordConfirmation"
          autoComplete="new-email"
          containerClassName="my-4"
          className="mr-1 mt-1 mb-2 lg:mt-2 bg-blue-200 text-secondary placeholder:text-secondary font-semibold rounded-full input-secondary border-none invalid:text-red-500 autofill:text-secondary autofill:bg-blue-200"
          testId="passwordConfirmation-error-message"
          showPassword={true}
        />

        <Submit data-testid="forgotPasswordBtn" className="rounded-full text-md" title={isLoading ? <Loader /> : "Change Password"} />
        <div className="divider">OR</div>
        <Link className="text-primary font-bold mb-10 text-center" to={routesPath.REGISTER}>
          Create New Account
        </Link>
        <Link className="text-primary font-bold mt-10 text-center" to={routesPath.LOGIN}>
          Back to Login
        </Link>
      </Form>
    </>
  );
};

export default FormResetPassword;
