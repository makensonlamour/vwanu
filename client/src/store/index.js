import { configureStore } from "@reduxjs/toolkit";
import thunkMiddleware from "redux-thunk";
import reducer from "./reducer";
import api from "../middleware/api";
import alerts from "../middleware/alerts";

export const store = configureStore({
  reducer,
  middleware: [thunkMiddleware, api, alerts],
});
