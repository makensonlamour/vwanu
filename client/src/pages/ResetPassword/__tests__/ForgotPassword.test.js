import React from "react";
import { render as rtlRender, screen, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import { act as testAct, create } from "react-test-renderer";
import { act as domAct } from "react-dom/test-utils";
import { store } from "../../../store";
import ForgotPassword from "../ForgotPassword";
import { Provider } from "react-redux";

afterEach(cleanup);

const render = (component) => rtlRender(<Provider store={store}>{component}</Provider>);

//Testing ForgotPassword form
//Test Dom Elements
describe("ForgotPassword Test Dom Elements", () => {
  it("match snapshot", () => {
    let tree;
    domAct(() => {
      testAct(() => {
        tree = create(
          <Provider store={store}>
            <Router>
              <ForgotPassword />
            </Router>
          </Provider>
        );
      });
    });
    expect(tree).toMatchSnapshot();
  });
});

it("should have a form", () => {
  render(
    <Router>
      <ForgotPassword />
    </Router>
  );

  const ForgotPasswordForm = screen.getByRole("form");
  expect(ForgotPasswordForm).toBeInTheDocument();
});

it("should have an email input", () => {
  render(
    <Router>
      <ForgotPassword />
    </Router>
  );

  const emailInput = screen.getByPlaceholderText("Email");
  expect(emailInput).toBeInTheDocument();
});

it("should have a Reset Password button", () => {
  render(
    <Router>
      <ForgotPassword />
    </Router>
  );

  const submitButton = screen.getByRole("button", { name: "Reset Password" });
  expect(submitButton).toBeInTheDocument();
});

it("the submit button should have the text Reset Password", () => {
  render(
    <Router>
      <ForgotPassword />
    </Router>
  );

  const submitButton = screen.getByRole("button", { name: "Reset Password" });
  expect(submitButton).toHaveTextContent("Reset Password");
});

//Test require fields
describe("Test require fields", () => {
  it("the email input should be a required fields", () => {
    render(
      <Router>
        <ForgotPassword />
      </Router>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    expect(emailInput).toBeInTheDocument();
    expect(emailInput).toHaveAttribute("required");
  });

  it("on initial render, the submit button is enabled", () => {
    render(
      <Router>
        <ForgotPassword />
      </Router>
    );

    const submitButton = screen.getByRole("button", { name: "Reset Password" });
    expect(submitButton).not.toBeDisabled();
  });
});

//test ForgotPassword functionality
describe("Test ForgotPassword Functionality", () => {
  it(" should show error message email when the form is submitted with empty email", async () => {
    domAct(() => {
      render(
        <Router>
          <ForgotPassword />
        </Router>
      );
    });

    const emailInput = screen.getByPlaceholderText("Email");
    const submitButton = screen.getByRole("button", { name: "Reset Password" });

    fireEvent.change(emailInput, { target: { value: "" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("email-error-message")).toHaveTextContent("Email is a required field"));
  });

  it(" should show error message email if email missing @", async () => {
    domAct(() => {
      render(
        <Router>
          <ForgotPassword />
        </Router>
      );
    });

    const emailInput = screen.getByPlaceholderText("Email");
    const submitButton = screen.getByRole("button", { name: "Reset Password" });

    fireEvent.change(emailInput, { target: { value: "testtest.com" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("email-error-message")).toHaveTextContent("Email must be a valid email"));
  });

  it(" should show error message email if email missing .", async () => {
    domAct(() => {
      render(
        <Router>
          <ForgotPassword />
        </Router>
      );
    });

    const emailInput = screen.getByPlaceholderText("Email");
    const submitButton = screen.getByRole("button", { name: "Reset Password" });

    fireEvent.change(emailInput, { target: { value: "test@testcom" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("email-error-message")).toHaveTextContent("Email must be a valid email"));
  });

  it(" shouldn't display error", async () => {
    domAct(() => {
      render(
        <Router>
          <ForgotPassword />
        </Router>
      );
    });

    const emailInput = screen.getByPlaceholderText("Email");
    const submitButton = screen.getByRole("button", { name: "Reset Password" });

    fireEvent.change(emailInput, { target: { value: "test@test.com" } });

    submitButton.click();
    await waitFor(() => expect(screen.queryByTestId("email-error-message")).toBeNull());
  });
});
