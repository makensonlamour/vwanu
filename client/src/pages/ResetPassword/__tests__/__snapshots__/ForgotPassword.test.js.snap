// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword Test Dom Elements match snapshot 1`] = `
<div
  className=""
>
  <div
    className="mt-10 mx-10"
  >
    <div
      className="mt-4 mb-6 lg:mb-10"
    >
      <img
        alt="logo_vwanu"
        className="w-2/5 lg:w-1/5 m-auto"
        src="Asset_2.png"
      />
    </div>
    <div
      className="justify-center m-auto md:px-16 lg:px-0 md:w-2/3 lg:w-1/3"
    >
      <div
        className="px-5 lg:px-10 pt-5 pb-10 card shadow-lg rounded-3xl my-10"
        role="form"
      >
        <h1
          className="card-title text-secondary text-md text-center"
        >
          Forgot your password?
        </h1>
        <p
          className="text-sky-600"
        >
          Enter your email and we'll send you a link to reset your password.
        </p>
        <div
          className="form-control"
        >
          <label
            className="label"
          >
            <span
              className="label-text"
            />
          </label>
          <input
            autoCapitalize="none"
            autoComplete="new-email"
            className="input mt-1 lg:mt-2 bg-blue-200 text-secondary placeholder:text-secondary font-semibold rounded-full input-secondary border-none invalid:text-error autofill:text-secondary autofill:bg-blue-200"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Email"
            required={true}
            type="email"
            value=""
          />
        </div>
        <button
          className="btn btn-primary mt-4 normal-case text-base-100 hover:bg-secondary rounded-full text-md"
          data-testid="forgotPasswordBtn"
          onClick={[Function]}
          type="submit"
        >
          Reset Password
        </button>
        <div
          className="divider"
        >
          OR
        </div>
        <a
          className="text-primary font-bold mb-10 text-center"
          href="/register"
          onClick={[Function]}
        >
          Create New Account
        </a>
        <a
          className="text-primary font-bold mt-10 text-center"
          href="/login"
          onClick={[Function]}
        >
          Back to Login
        </a>
      </div>
    </div>
  </div>
</div>
`;
